test/e2e/network_segmentation_services.go

```diff
@@ -0,0 +1,376 @@
+					func(pod *v1.Pod) {
+						pod.Spec.Containers[0].Ports = []v1.ContainerPort{{ContainerPort: (serviceTargetPort), Protocol: "UDP"}}
+					})
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+				ginkgo.By("create a node port service in the default network")
+				defaultService := &v1.Service{
+					ObjectMeta: metav1.ObjectMeta{Name: "service-default"},
+					Spec: v1.ServiceSpec{
+						Ports: []v1.ServicePort{
+							{
+								Name:     "udp-port",
+								Port:     int32(servicePort),
+								Protocol: v1.ProtocolUDP,
+							},
+						},
+						Selector:       defaultLabels,
+						Type:           v1.ServiceTypeNodePort,
+						IPFamilyPolicy: &policy,
+					},
+				}
+
+				defaultService, err = f.ClientSet.CoreV1().Services(defaultNetNamespace).Create(context.TODO(), defaultService, metav1.CreateOptions{})
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+				ginkgo.By("verify that the client pod in the UDN cannot connect to the default-network service")
+				checkNoConnectionToClusterIPs(f, udnClientPod2, defaultServerPod, defaultService)
+				checkNoConnectionToNodePort(f, udnClientPod2, defaultService, &nodes.Items[0], "server node")
+				checkNoConnectionToNodePort(f, udnClientPod2, defaultService, &nodes.Items[1], "same node")
+				checkNoConnectionToNodePort(f, udnClientPod2, defaultService, &nodes.Items[2], "other node")
+
+			},
+
+			ginkgo.Entry(
+				"L3 dualstack primary UDN, cluster-networked pods, NodePort service",
+				networkAttachmentConfigParams{
+					name:     nadName,
+					topology: "layer3",
+					cidr:     fmt.Sprintf("%s,%s", userDefinedNetworkIPv4Subnet, userDefinedNetworkIPv6Subnet),
+					role:     "primary",
+				},
+			),
+
+			// TODO add L2 NADs once L2 UDN support is available for services
+		)
+
+	})
+
+})
+
+// TODO Once https://github.com/ovn-org/ovn-kubernetes/pull/4567 merges, use the vendored *TestJig.Run(), which tests
+// the reachability of a service through its name and through its cluster IP. For now only test the cluster IP.
+
+const OVNNodeHostCIDRs = "k8s.ovn.org/host-cidrs"
+
+// ParseNodeHostCIDRsDropNetMask returns the parsed host IP addresses found on a node's host CIDR annotation. Removes the mask.
+func ParseNodeHostCIDRsDropNetMask(node *kapi.Node) (sets.Set[string], error) {
+	addrAnnotation, ok := node.Annotations[OVNNodeHostCIDRs]
+	if !ok {
+		return nil, newAnnotationNotSetError("%s annotation not found for node %q", OVNNodeHostCIDRs, node.Name)
+	}
+
+	var cfg []string
+	if err := json.Unmarshal([]byte(addrAnnotation), &cfg); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal host cidrs annotation %s for node %q: %v",
+			addrAnnotation, node.Name, err)
+	}
+
+	for i, cidr := range cfg {
+		ip, _, err := net.ParseCIDR(cidr)
+		if err != nil || ip == nil {
+			return nil, fmt.Errorf("failed to parse node host cidr: %v", err)
+		}
+		cfg[i] = ip.String()
+	}
+	return sets.New(cfg...), nil
+}
```

Comment:
This will also include the secondary node IPs (if multi-net is enabled - which is the case with UDN).  Connectiong NodePort services on remote node secondary IPs doesn't work.  That's the case for the default network too.

Would it make sense to use the `"k8s.ovn.org/node-primary-ifaddr"` annotation instead?  Something like (not fully tested):
```
// ParseNodeHostIPDropNetMask returns the parsed host IP addresses found on a node's host CIDR annotation. Removes the mask.
func ParseNodeHostIPDropNetMask(node *kapi.Node) (sets.Set[string], error) {
	nodeIfAddrAnnotation, ok := node.Annotations[OvnNodeIfAddr]
	if !ok {
		return nil, newAnnotationNotSetError("%s annotation not found for node %q", OvnNodeIfAddr, node.Name)
	}
	nodeIfAddr := &primaryIfAddrAnnotation{}
	if err := json.Unmarshal([]byte(nodeIfAddrAnnotation), nodeIfAddr); err != nil {
		return nil, fmt.Errorf("failed to unmarshal annotation: %s for node %q, err: %v", OvnNodeIfAddr, node.Name, err)
	}

	var cfg []string
	if nodeIfAddr.IPv4 != "" {
		cfg = append(cfg, nodeIfAddr.IPv4)
	}
	if nodeIfAddr.IPv6 != "" {
		cfg = append(cfg, nodeIfAddr.IPv6)
	}
	if len(cfg) == 0 {
		return nil, fmt.Errorf("node: %q does not have any IP information set", node.Name)
	}

	for i, cidr := range cfg {
		ip, _, err := net.ParseCIDR(cidr)
		if err != nil || ip == nil {
			return nil, fmt.Errorf("failed to parse node host cidr: %v", err)
		}
		cfg[i] = ip.String()
	}
	return sets.New(cfg...), nil
}
```